### PR TITLE
fix(amf): Support for dotted dnn encoding and decoding

### PR DESCRIPTION
**Pull request by Sathyaj27**
Signed-off-by: Sathyaj27 <sathya.jayadev@wavelabs.ai>

## Summary

Dotted DNN support is added.
Decoding and encoding added and UT added for the same.

## Test Plan

1) Tested positive case for dotted dnn
MicrosoftTeams-image
2) Tested negative  case of dotted dnn
MicrosoftTeams-image (1)
3) Maximum DNN length 100 
MicrosoftTeams-image (2)
lengthmore

4)Tested existing cases.
## Additional Information

- [ ] This change is backwards-breaking
Corresponding zenhub task id #10603

### CODE REVIEW

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 1]
```diff
@@ -928,6 +928,56 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
...
 }
 
+TEST(test_optional_dnn_dotted_pdu, test_pdu_session_establish_optional) {
+  uint32_t bytes         = 0;
+  uint32_t container_len = 0;
+  bstring buffer;
+  amf_nas_message_t msg = {};
+
+  // build uplinknastransport
+  // uplink nas transport(pdu session request)
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: remove if unused

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 2]
```diff
@@ -68,15 +82,28 @@ int DNNMsg::EncodeDNNMsg(
...
+    if (dnn_len + dnn_length < dnn_message->len - 1) {
+      if (memcmp(dnn_message->dnn + dnn_length + dnn_len, ".", 1) == 0)
+        dnn_length = dnn_length + 1;
+    }
+    dnn_length = dnn_length + dnn_len;
+
+    BUFFER_PRINT_LOG(buffer + encoded, dnn_len);
+    MLOG(MDEBUG) << "dnn str : " << dnn_message->dnn;
+    encoded = encoded + dnn_len;
+  }
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: which test covers these?

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 3]
```diff
@@ -68,15 +82,28 @@ int DNNMsg::EncodeDNNMsg(
...
-  ENCODE_U8(buffer + encoded, dnn_message->len - 1, encoded);
-  MLOG(MDEBUG) << "dnn len : " << dnn_message->len - 1;
-
-  memcpy(buffer + encoded, dnn_message->dnn, dnn_message->len - 1);
-  BUFFER_PRINT_LOG(buffer + encoded, dnn_message->len - 1);
-  MLOG(MDEBUG) << "dnn str : " << dnn_message->dnn;
-  encoded = encoded + dnn_message->len - 1;
-
+  uint8_t dnn_length = 0;
+  while (dnn_length < dnn_message->len - 1) {
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp:83]: why is this a while loop?

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 4]
```diff
@@ -39,15 +39,29 @@ int DNNMsg::DecodeDNNMsg(
...
   DECODE_U8(buffer + decoded, dnn_len, decoded);
   MLOG(MDEBUG) << "dnn_len : " << static_cast<int>(dnn_len);
 
   memcpy(dnn_message->dnn, buffer + decoded, dnn_len);
-
+  dnn_length += dnn_len;
   decoded = decoded + dnn_len;
-  MLOG(MDEBUG) << "dnn str : " << dnn_message->dnn;
 
+  while (dnn_length + 1 < ielen) {
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp:51]: why is this a while loop? how is the dot separated dnn is encoded as IE?

Comment (Sathyaj27) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp:51]: Hi Ulas,
The dotted DNN is in the format label1.label2...label- n.
Its encoded in the format (total length, label1 length , label1, label2 length, label 2, ....label-n length, label-n)
dottendnn

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp (hunk 2 above)
Comment (Sathyaj27) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: test_optional_dnn_dotted_pdu covers both encoding and decoding for dnn with dotted format.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp (hunk 1 above)
Comment (Sathyaj27) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:938]: Ulas,
Added this test case for UT of dotted dnn encoding and decoding

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp (hunk 4 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp:51]: thanks for the clarification.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 5]
```diff
@@ -39,15 +39,29 @@ int DNNMsg::DecodeDNNMsg(
...
+    uint8_t dnn_len = 0;
+    memcpy(dnn_message->dnn + dnn_length, ".", 1);
+    DECODE_U8(buffer + decoded, dnn_len, decoded);
+    dnn_length = dnn_length + 1;
+    MLOG(MDEBUG) << "dnn_len : " << static_cast<int>(dnn_len);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: no need for this.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 6]
```diff
@@ -39,15 +39,29 @@ int DNNMsg::DecodeDNNMsg(
...
+
+    memcpy(dnn_message->dnn + dnn_length, buffer + decoded, dnn_len);
+
+    decoded = decoded + dnn_len;
+    dnn_length += dnn_len;
+    MLOG(MDEBUG) << "dnn: " << dnn_message->dnn;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: no need for this.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 7]
```diff
@@ -68,15 +82,28 @@ int DNNMsg::EncodeDNNMsg(
...
+  uint8_t dnn_length = 0;
+  while (dnn_length < dnn_message->len - 1) {
+    uint8_t dnn_len = 0;
+    for (int i = dnn_length; (memcmp(dnn_message->dnn + i, ".", 1) != 0) &&
+                             i < dnn_message->len - 1;
+         i++) {
+      ++dnn_len;
+    }
+    ENCODE_U8(buffer + encoded, dnn_len, encoded);
+    MLOG(MDEBUG) << "dnn len : " << dnn_len;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: no need for this log.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 8]
```diff
@@ -68,15 +82,28 @@ int DNNMsg::EncodeDNNMsg(
 
   ENCODE_U8(buffer + encoded, dnn_message->len, encoded);
   MLOG(MDEBUG) << "len : " << static_cast<int>(dnn_message->len);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: no need for this.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp [hunk 9]
```diff
@@ -68,15 +82,28 @@ int DNNMsg::EncodeDNNMsg(
...
+
+    memcpy(buffer + encoded, dnn_message->dnn + dnn_length, dnn_len);
+    if (dnn_len + dnn_length < dnn_message->len - 1) {
+      if (memcmp(dnn_message->dnn + dnn_length + dnn_len, ".", 1) == 0)
+        dnn_length = dnn_length + 1;
+    }
+    dnn_length = dnn_length + dnn_len;
+
+    BUFFER_PRINT_LOG(buffer + encoded, dnn_len);
+    MLOG(MDEBUG) << "dnn str : " << dnn_message->dnn;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: no need to add log strings for debugging functional correctness. unit tests should take care of this.

Comment (Sathyaj27) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp]: Have removed all debugging logs.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp (hunk 1 above)
Comment (Sathyaj27) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: removed unused code

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**